File: test/test_field_copy.cpp
#include <catch2/catch_test_macros.hpp>
#include <catch2/generators/catch_generators.hpp>
#include <Omega_h_mesh.hpp>
#include <Omega_h_build.hpp>
#include <Omega_h_for.hpp>
#include "pcms/transfer/copy.h"
#include "pcms/field/function_space/lagrange.h"
#include "pcms/field/field_metadata.h"
#include "pcms/utility/assert.h"
#include <Kokkos_Core.hpp>

using pcms::Real;

void test_copy(Omega_h::CommPtr world, int dim, int order, int num_components)
{
  int nx = 100;
  int ny = dim > 1 ? 100 : 0;
  int nz = dim > 2 ? 100 : 0;

  auto mesh =
    Omega_h::build_box(world, OMEGA_H_SIMPLEX, 1, 1, 1, nx, ny, nz, false);
  auto factory = pcms::LagrangeFunctionSpace::FromMesh(
    mesh, order, num_components, pcms::CoordinateSystem::Cartesian, "global",
    pcms::LagrangeFunctionSpace::Backend::OmegaH);
  auto layout = factory.GetLayout();
  int ndata = layout->GetNumOwnedDofHolder() * num_components;
  Omega_h::HostWrite<Real> ids(ndata);
  Kokkos::parallel_for(
    Kokkos::RangePolicy<pcms::HostMemorySpace::execution_space>(0, ndata),
    [=](int i) { ids[i] = i; });

  auto original = factory.CreateField<Real>(pcms::FieldMetadata{});
  original.SetDOFHolderDataHost(
    pcms::Rank2View<const Real, pcms::HostMemorySpace>(
      ids.data(), layout->GetNumOwnedDofHolder(), num_components));

  auto copied = factory.CreateField<Real>(pcms::FieldMetadata{});
  pcms::Copy<Real> copy(factory, factory);
  copy.Apply(original, copied);
  auto copied_array = pcms::FlattenToRank1View(copied.GetDOFHolderDataHost());

  REQUIRE(copied_array.size() == ndata);
  int sum = 0;
  Kokkos::parallel_reduce(
    Kokkos::RangePolicy<pcms::HostMemorySpace::execution_space>(0, ndata),
    [=](int i, int& local_sum) {
      local_sum += std::abs(ids[i] - copied_array[i]) < 1e-12;
    },
    sum);
  REQUIRE(sum == ndata);
}

TEST_CASE("copy omega_h_field2 data")
{
  auto lib = Omega_h::Library{};
  test_copy(lib.world(), 2, 0, 1);
  test_copy(lib.world(), 2, 1, 1);
  auto mesh = Omega_h::build_box(lib.world(), OMEGA_H_SIMPLEX, 1, 1, 1, 100,
                                 100, 0, false);
  REQUIRE_THROWS_AS(pcms::LagrangeFunctionSpace::FromMesh(
                      mesh, 2, 1, pcms::CoordinateSystem::Cartesian, "global",
                      pcms::LagrangeFunctionSpace::Backend::OmegaH),
                    pcms::pcms_error);
}

File: test/python/test_field_copy.py
"""Test copying Omega_h field data."""

import numpy as np
import pytest
import pcms
import PyOmega_h as omega_h


class TestFieldCopy:
    """Tests for field data copying between Omega_h-backed fields."""

    @pytest.mark.parametrize("dim, order, num_components", [
        (2, 1, 1),
        (2, 2, 1),
    ])
    def test_copy(self, world, dim, order, num_components):
        """Test copying omega_h field data."""
        nx = 100
        ny = 100 if dim > 1 else 0
        nz = 100 if dim > 2 else 0

        # Build mesh
        mesh = omega_h.build_box(
            world,
            omega_h.Family.SIMPLEX,
            1.0, 1.0, 1.0,
            nx, ny, nz,
            False
        )

        # Create factory and layout
        factory = pcms.LagrangeFunctionSpace.from_mesh(
            mesh, order, num_components, pcms.CoordinateSystem.Cartesian
        )

        # Create original field and set data
        original = factory.create_field()
        ndata = original.get_num_dof_holders() * original.get_num_components()

        # Create sequential array of IDs
        ids = np.arange(ndata, dtype=np.float64)
        original.set_dof_holder_data(ids)

        # Create copied field and copy data
        copied = factory.create_field()
        copier = pcms.Copy(factory, factory)
        copier.apply(original, copied)

        # Get copied data
        copied_array = copied.get_dof_holder_data()

        # Verify the copy
        assert len(copied_array) == ndata, \
            f"Expected {ndata} elements, got {len(copied_array)}"

        # Check that all values match
        differences = np.abs(ids - copied_array)
        num_matches = np.sum(differences < 1e-12)
        assert num_matches == ndata, \
            f"Only {num_matches}/{ndata} elements matched"
